File: tests/fixtures/safe/secure-code.ts
// Safe: Parameterized SQL queries
import { Pool } from 'pg';

const pool = new Pool();

async function getUser(userId: string) {
  const result = await pool.query("SELECT * FROM users WHERE id = $1", [userId]);
  return result.rows[0];
}

async function searchUsers(name: string) {
  const result = await pool.query("SELECT * FROM users WHERE name = $1", [name]);
  return result.rows;
}

// Safe: No dynamic command execution
import { execSync } from 'child_process';

function runFixedCommand() {
  execSync("ls -la /tmp");
}

// Safe: No eval or dynamic code
function add(a: number, b: number) {
  return a + b;
}

function greet(name: string) {
  return `Hello, ${name}!`;
}

// Safe: no nested/overlapping quantifiers
function isValidEmail(input: string) {
  return new RegExp("^[a-zA-Z0-9]+@").test(input);
}

// Safe: querying by a specific validated field, not the whole request body
async function login(users: any, username: string) {
  return users.findOne({ username });
}

// Safe: redirect target is a fixed, known path
function goToLogin(res: any) {
  res.redirect("/login");
}

// Safe: restricted to a specific signing algorithm
function verifyToken(jwt: any, token: string, secret: string) {
  return jwt.verify(token, secret, { algorithms: ['HS256'] });
}

// Safe: entity substitution disabled
function parseXml(libxmljs: any, data: string) {
  return libxmljs.parseXml(data, { noent: false });
}

File: tests/fixtures/safe/secure-code.py
# Safe: Parameterized queries
import sqlite3

def get_user(user_id):
    conn = sqlite3.connect('app.db')
    cursor = conn.cursor()
    cursor.execute("SELECT * FROM users WHERE id = ?", (user_id,))
    return cursor.fetchone()

# Safe: No shell injection
import subprocess

def ping_host(host):
    subprocess.run(["ping", "-c", "4", host], check=True)

# Safe: Using safe loader
import yaml

def load_config(config_str):
    return yaml.safe_load(config_str)

# Safe: Normal function
def add(a, b):
    return a + b

# Safe: mark_safe on a fixed literal, not attacker-controlled input
from django.utils.safestring import mark_safe

def render_banner():
    return mark_safe("<b>Welcome</b>")

# Safe: cryptographic RNG for a reset token
import secrets

def generate_password_reset_token():
    return secrets.token_hex(16)

# Safe: no nested/overlapping quantifiers
import re

def is_valid_email(value):
    return re.compile(r"^[a-zA-Z0-9]+@").match(value)

# Safe: querying by a specific validated field, not the whole request body
def login(users, username):
    return users.find_one({"username": username})

# Safe: redirect target is a fixed, known path
def go_to_login():
    return redirect("/login")

# Safe: restricted to a specific signing algorithm
def verify_token(jwt, token, key):
    return jwt.decode(token, key, algorithms=["HS256"])

# Safe: external entities not resolved, no network access
def parse_xml(etree, data):
    parser = etree.XMLParser(resolve_entities=False, no_network=True)
    return etree.fromstring(data, parser)
